File: API/Entities/Basket.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace API.Entities
{
    public class Basket
    {
        public int Id { get; set; }
        public required string BasketId { get; set; }
        public List<BasketItem> Items { get; set; } = new();
        public string? ClientSecret { get; set; }
        public string? PaymentIntentId { get; set; }

        public void AddItem(Product product, int quantity)
        {
            if(product == null) ArgumentNullException.ThrowIfNull(product);
            if(quantity <= 0) throw new ArgumentException("Quantity must be greater than zero.", nameof(quantity));

            var existingItem = FindProduct(product.Id);

            if(existingItem == null)
            {
                Items.Add(new BasketItem 
                {
                    Quantity = quantity,
                    Product = product
                });
            } 
            else 
            {
                existingItem.Quantity += quantity;
            }
            

        }

        public void RemoveItem(int productId, int quantity)
        {
            if(quantity <= 0) throw new ArgumentException("Quantity must be greater than zero.", nameof(quantity));

            var existingItem = FindProduct(productId);
            if(existingItem == null) return;

            existingItem.Quantity -= quantity;
            if(existingItem. Quantity <= 0) Items.Remove(existingItem);

        }

        //Find product helper function
        private BasketItem? FindProduct(int id)
        {
            return Items.FirstOrDefault(x => x.ProductId == id);
        }
    }
}

File: client/src/app/models/basket.ts
import { type Product } from "./product";

export type Basket = {
  id: number;
  basketId: string;
  items: Item[];
  clientSecret?: string;
  paymentIntentId?: string;
};

export class Item {
  constructor(product: Product, quantity: number) {
    this.productId = product.id;
    this.name = product.name;
    this.price = product.price;
    this.imageUrl = product.imageUrl;
    this.category = product.category;
    this.band = product.band;
    this.quantity = quantity;
  }

  productId: number;
  name: string;
  price: number;
  imageUrl: string;
  category: string;
  band: string;
  quantity: number;
}
